caffe2/CMakeLists.txt

```diff
@@ -126,11 +126,9 @@ if (BUILD_TEST)
     # libraries even if they are not part of the public dependency libs. As a
     # result, we will explicitly link the test against the Caffe2 dependency
     # libs.
-    target_link_libraries(${test_name} ${Caffe2_MAIN_LIBS})
-    target_link_libraries(${test_name} gtest_main)
-    target_link_libraries(${test_name} ${Caffe2_DEPENDENCY_LIBS})
+    target_link_libraries(${test_name} ${Caffe2_MAIN_LIBS} gtest_main)
```

Comment:
The left lines are actually all needed - see comments above - could you add them back?

Reply:
The reason to remove them is that with them I saw `${Caffe2_MAIN_LIBS}` placed **after** `${Caffe2_DEPENDENCY_LIBS}` so that ld cannot resolve some symbols although `${Caffe2_MAIN_LIBS}` was added before `${Caffe2_DEPENDENCY_LIBS} . With them, I saw messages like:
```
[ 82%] Linking CXX executable ../bin/workspace_test
cd /hack/freedom/ml/caffe2/build_android/caffe2 && /tmp/cmake-3.10.2-Linux-x86_64/bin/cmake -E cmake_link_script CMakeFiles/workspace_test.dir/link.txt --verbose=1
/home/freedom/work/android-ndk-r16/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/bin/arm-linux-androideabi-g++ --sysroot=/home/freedom/work/android-ndk-r16/sysroot  -isystem /home/freedom/work/android-ndk-r16/sysroot/usr/include/arm-linux-androideabi -D__ANDROID_API__=21 -g -DANDROID -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -march=armv7-a -mfloat-abi=softfp -mfpu=vfpv3-d16 -mthumb -Wa,--noexecstack -Wformat -Werror=format-security -frtti -fexceptions -s -Wno-deprecated -std=c++11 -O2 -fPIC -Wno-narrowing -Wno-invalid-partial-specialization -Os -DNDEBUG   -Wl,--exclude-libs,libgcc.a -Wl,--exclude-libs,libatomic.a --sysroot /home/freedom/work/android-ndk-r16/platforms/android-21/arch-arm -Wl,--build-id -Wl,--warn-shared-textrel -Wl,--fatal-warnings -Wl,--fix-cortex-a8 -Wl,--no-undefined -Wl,-z,noexecstack -Wl,-z,relro -Wl,-z,now -Wl,--gc-sections -Wl,-z,nocopyreloc -pie -fPIE  CMakeFiles/workspace_test.dir/core/workspace_test.cc.o  -o ../bin/workspace_test ../lib/libgtest_main.a ../lib/libnnpack.a ../lib/libcpufeatures.a -llog -ldl ../lib/libcpuinfo.a ../lib/libpthreadpool.a -Wl,--whole-archive /hack/freedom/ml/caffe2/build_android/lib/libcaffe2.a -Wl,--no-whole-archive ../lib/libprotobuf.a /home/freedom/work/android-ndk-r16/platforms/android-21/arch-arm/usr/lib/libz.so ../lib/libgtest.a -latomic -lm "/home/freedom/work/android-ndk-r16/sources/cxx-stl/gnu-libstdc++/4.9/libs/armeabi-v7a/libgnustl_static.a" 
/hack/freedom/ml/caffe2/caffe2/share/contrib/nnpack/conv_op.cc:36: error: undefined reference to 'nnp_initialize'
/hack/freedom/ml/caffe2/caffe2/share/contrib/nnpack/conv_op.cc:218: error: undefined reference to 'nnp_convolution_inference'
/hack/freedom/ml/caffe2/caffe2/share/contrib/nnpack/conv_op.cc:254: error: undefined reference to 'nnp_convolution_inference'
/hack/freedom/ml/caffe2/caffe2/share/contrib/nnpack/conv_op.cc:317: error: undefined reference to 'nnp_convolution_inference'
/hack/freedom/ml/caffe2/caffe2/share/contrib/nnpack/conv_op.cc:338: error: undefined reference to 'nnp_convolution_inference'
/hack/freedom/ml/caffe2/caffe2/utils/threadpool/ThreadPool.cc:53: error: undefined reference to 'android_getCpuCount'
/hack/freedom/ml/caffe2/caffe2/utils/threadpool/ThreadPool.cc:70: error: undefined reference to 'android_getCpuIdArm'
collect2: error: ld returned 1 exit status
make[2]: *** [bin/workspace_test] Error 1
make[2]: Leaving directory `/hack/freedom/ml/caffe2/build_android'
make[1]: *** [caffe2/CMakeFiles/workspace_test.dir/all] Error 2
make[1]: Leaving directory `/hack/freedom/ml/caffe2/build_android'
make: *** [all] Error 2
```

With them removed, I can build with ` -DBUILD_TEST=ON -DBUILD_BINARY=ON` successfully.

Reply:
@Yangqing I'm in favor of linking only `${Caffe2_MAIN_LIBS}` and let CMake sort out other dependencies

Reply:
I get the same error when building caffe2_benchmark. Can we quickly resolve the issues and merge the PR so that the regression can move on? @Yangqing @Maratyszcza 

Thanks.